### DISCUSSION TITLE
Is there a way to control available plugin in CMS plugin dropdown from text editor

### DISCUSSION THREAD

**Discussion by sveetch** [django-cms/django-cms#7721] [Q&A] [answered]
Actually, a text plugin like djangocms-text-ckeditor can have a dropdown menu to  select a [plugin allowed for child plugin in text editor](https://docs.django-cms.org/en/latest/reference/plugins.html?highlight=text_enabled#cms.plugin_base.CMSPluginBase.text_enabled) .

This is a nice feature but it may be a problem because it is the plugin itself which determine if it will be in this list of allowed child plugins.

For example, i have a project with some external plugins and some of them are not really implemented in the project webdesign to be included in text content. This is not very good for the user experience that may not know about this limitation and he would not understand  why it can include a plugin that is available from this dropdown. 

But i did not found any way to control this list from a DjangoCMS or djangocms-text-ckeditor setting. I thought maybe there was a way to do it in `CMS_PLACEHOLDER_CONF` but there is not. From what i looked in `plugin_base` and `plugin_pool`, i did not found anything that could overwrite it but i may be wrong.

So, is there a trick to disable some plugins with `text_enabled=True` to not be in the list of child plugin candidates in text plugin ?

**Comment by fsbraun** [comment, marked as answer]
Hi @sveetch ! Thank you for bringing this up!

As a general note, unused plugins should be removed from the project. This can be done either by removing its app from the installed apps or by actively unregistering the plugin in your project (e.g., if you need other plugins from that app).

Regarding text fields, there are three opinions around:
1. **Do not allow any child plugins in text fields.** 
   This would make HTML fields and Text plugins equivalent. It would, however, need a way to enter dynamic links into the HTML potentially for both the `HTMLField` and the `TextPlugin`. Other elements such as images would never go into a rich text field.
2. **Make it configurable which plugins I can have inside a text field**
   This is what you are looking for. It has two approaches to it: Either add some kind of settings to djangocms-text-ckeditor to exclude plugins from being listed, or - also for user experience - have child plugins as separate icons of the CKEditor toolbar which can already be configured.
3. **Provide a consistent selection of plugins inside text fields and outside.**
   This is the current situation. The two issues are: the difference between `HTMLFields` and `TextPlugin`, usability with the pull down menu in the editor, lack of configurability.

**The opinions I hear tend towards (1) and (2). I am happy to use this discussion to also set the stage for a future replacement of djangocms-text-ckeditor by a general Rich Text Editor plugin, potentially with multiple frontends for editors such as CKEditor or Tiptap. So, please let me know your thoughts!**

@sveetch Since django CMS is a community-driven project, I encourage you to propose a black list setting for text-plugins as a quick fix to be included in djangocms-text-ckeditor. The place to remove the plugins from the pulldown menu probably is here: 

https://github.com/django-cms/djangocms-text-ckeditor/blob/fea95ed16c4eff32c52686aec59a48ea7971a3ff/djangocms_text_ckeditor/cms_plugins.py#L496-L503

**Comment by sveetch** [comment]
> As a general note, unused plugins should be removed from the project. This can be done either by removing its app from the installed apps or by actively unregistering the plugin in your project (e.g., if you need other plugins from that app).

Yeah but my problem is that these plugins are effectively used as child plugins but not wanted as text plugin child inside text content itself. Concretely i would like to enable including image with filer picture plugin, but don't want to support filer folder or plugins (like i have django-fobi that provide a plugin to include a form that we definitively do not want inside text).

**Comment by sveetch** [comment]
> Regarding text fields, there are three opinions around:
> 
>     1. **Do not allow any child plugins in text fields.**
>        This would make HTML fields and Text plugins equivalent. It would, however, need a way to enter dynamic links into the HTML potentially for both the `HTMLField` and the `TextPlugin`. Other elements such as images would never go into a rich text field.

By the way, HTMLField does not support child plugin inside text plugin since plugin need a placeholder, djangocms-text-ckeditor documentation recommended to use a PlaceholderField instead of an HtmlField for this usecase. So finally the point with cms plugin dropdown is only for TextPlugin.

>     2. **Make it configurable which plugins I can have inside a text field**
>        This is what you are looking for. It has two approaches to it: Either add some kind of settings to djangocms-text-ckeditor to exclude plugins from being listed, or - also for user experience - have child plugins as separate icons of the CKEditor toolbar which can already be configured.

Yep this is the way for me, either an optional blacklist or even a whitelist from a setting. I don't know however if it is best in djangocms-text-ckeditor or CMS itself, i found CMS was providing the data to fill the dropdown for text plugins that have to implement it with their editor, so i thought it was best suited in CMS but i may be wrong.

>     3. **Provide a consistent selection of plugins inside text fields and outside.**
>        This is the current situation. The two issues are: the difference between `HTMLFields` and `TextPlugin`, usability with the pull down menu in the editor, lack of configurability.

I understood the technical problem with HtmlField widget and child plugins so i dropped this subject, at least if the TextPlugin is managing right the dropdown, this is fine for me.

> **The opinions I hear tend towards (1) and (2). I am happy to use this discussion to also set the stage for a future replacement of djangocms-text-ckeditor by a general Rich Text Editor plugin, potentially with multiple frontends for editors such as CKEditor or Tiptap. So, please let me know your thoughts!**

This is a very interesting subject you are introducing here since recently i started to study on CKEditor5 that is a great improvement but is not planned from either django-ckeditor or  djangocms-text-ckeditor. But to talk for the CMS stack, it may be something to work in CMS 4.x no ? I imagine there is a lot of work to do and it would be hard to port between 3.x and 4.x. But finally it seems a new subject,  i will follow it when it  comes to the scene.

> @sveetch Since django CMS is a community-driven project, I encourage you to propose a black list setting for text-plugins as a quick fix to be included in djangocms-text-ckeditor. The place to remove the plugins from the pulldown menu probably is here:

I can promise anything but it's something i could do, since it is a little bit easier to work on djangocms-text-ckeditor than on CMS.

**Reply by fsbraun** [reply]
The general idea of a rich text plugin is that you potentially could swap in your preferred editor. For now, django CMS 4 will work with djangocms-text-ckeditor.

Making changes to the djangocms-text-ckeditor seems right for me: the CMS offers the available text-enabled plugins, but the text plugin decides which it takes. It's a bit like configuring the ckeditor toolbar.

**Comment by sveetch** [comment]
Should this discussion be marked as answered since we ended on "There is currently no feature for this, we should implement it" ?

**Reply by fsbraun** [reply]
I'd love to get more feedback for a future rich text plugin.

**Comment by fsbraun** [comment]
It is configurable with https://github.com/django-cms/djangocms-text
